kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    // The commit file that this replay scanner will operate on.
+    commit_file: ParsedLogPath,
+    // The timestamp associated with this commit. By default this is the file modification time
+    // from the commit's [`FileMeta`]. If there is a [`CommitInfo`] with a timestamp generated by
+    // in-commit timestamps, that timestamp will be used instead.
+    //
+    // Note: This will be used once an expression is introduced to transform the engine data in
+    // [`TableChangesScanData`]
+    #[allow(unused)]
+    timestamp: i64,
+    // The data skipping filter for filtering log actions
+    filter: Option<DataSkippingFilter>,
+}
+
+fn prepare_table_changes(
+    commit_file: ParsedLogPath,
+    engine: &dyn Engine,
+    table_schema: &SchemaRef,
+    physical_predicate: Option<ExpressionRef>,
+) -> DeltaResult<LogReplayScanner> {
+    let filter = DataSkippingFilter::new(engine, table_schema, physical_predicate);
+    let visitor_schema = PreparePhaseVisitor::schema()?;
+    let action_iter = engine.get_json_handler().read_json_files(
+        &[commit_file.location.clone()],
+        visitor_schema,
+        None,
+    )?;
+
+    let mut remove_dvs = HashMap::default();
+    let mut add_paths = HashSet::new();
+    let mut has_cdc_action = false;
+    let mut timestamp = commit_file.location.last_modified;
+    for actions in action_iter {
+        let actions = actions?;
+        let selection_vector = match &filter {
+            Some(filter) => filter.apply(actions.as_ref())?,
+            None => vec![true; actions.len()],
+        };
+
+        let mut visitor = PreparePhaseVisitor::new(
+            selection_vector,
+            &mut add_paths,
+            &mut remove_dvs,
+            &mut has_cdc_action,
+            &mut timestamp,
+        );
+        visitor.visit_rows_of(actions.as_ref())?;
+
+        if let Some(protocol) = visitor.protocol {
+            protocol.ensure_read_supported()?;
+        }
+        if let Some((schema, configuration)) = visitor.metadata_info {
+            let schema: StructType = serde_json::from_str(&schema)?;
+            // Currently, schema compatibility is defined as having equal schema types. In the
+            // future, more permisive schema evolution will be supported.
+            // See: https://github.com/delta-io/delta-kernel-rs/issues/523
+            require!(
+                table_schema.as_ref() == &schema,
+                Error::change_data_feed_incompatible_schema(table_schema, &schema)
+            );
+            let table_properties = TableProperties::from(configuration);
+            require!(
+                table_properties.enable_change_data_feed.unwrap_or(false),
```

Comment:
@zachschuermann here's another case of cross-validation we need to add -- is it legal for the table property to exist if the table doesn't even include CDF in its protocol?

Reply:
yea ideally all of this will move to a new `TableConfiguration` like we talked about - I'm happy to tackle that soon as i get some time in the next few days